src/mlpack/methods/reinforcement_learning/environment/frozen_lake.hpp

```diff
@@ -0,0 +1,461 @@
+    {
+      Log::Info << "Episode terminated due to the maximum number of steps"
+          "being taken.\n";
+      return true;
+    }
+    else if (boardDescription[state.CurRow()][state.CurCol()] == 'H')
+    {
+      Log::Info << "Episode terminated due to agent falling in the hole.\n";
+      return true;
+    }
+    else if (boardDescription[state.CurRow()][state.CurCol()] == 'G')
+    {
+      Log::Info << "Episode terminated, agent reached the goal.\n";
+      return true;
+    }
+    return false;
+  }
+
+  //! Get the number of steps performed.
+  size_t StepsPerformed() const { return stepsPerformed; }
+
+  //! Get the maximum number of steps allowed.
+  size_t MaxSteps() const { return maxSteps; }
+
+  //! Set the maximum number of steps allowed.
+  size_t& MaxSteps() { return maxSteps; }
+
+  //! Get the board description.
+  std::vector<std::vector<char>> Description() const {return boardDescription;}
+
+  //! Set the board description.
+  std::vector<std::vector<char>>& Description() {return boardDescription;}
+
+ private:
+  /**
+   * Utility function that helps generate random environment board.
+   * 
+   * @return board, a 2D array of characters that hold the board description.
+   *    'S' is the Start Tile, the starting position of the agent.
+   *    'G' is the Goal Tile, the agent can walk into a goal tile to win.
+   *    'F' is a Frozen Tile, the agent can walk on it.
+   *    'H' is a Hole Tile, the agent can fall into it and the game ends.
+   */
+  std::vector<std::vector<char>> generateRandomBoard()
+  {
+    bool valid = false;
+    std::vector<std::vector<char>> Board;
+    //! TODO: Tentative, implement max step to control whether
+    // we can never generate a possible board.
+    while (!valid)
+    {
+      // Randomly choose tiles in the board (discrete random distribution).
+      Board = genBoardHelper(height, width, platformRate);
+      valid = dfsHelper(Board, height, width);
+    }
+    return Board;
+  }
+
+  /**
+   * Perform depth-firsth search to see if the environment board has 
+   * a solution or not.
+   * 
+   * @param candidateBoard Environment board description.
+   * @param height Height of the environment board.
+   * @param width Width of the environment board.
+   * @return true if there is a solution, false otherwise. 
+   */
+  static bool dfsHelper(std::vector<std::vector<char>> candidateBoard, size_t height, size_t width)
+  {
+    arma::Mat<short> visited(height, width);
+    visited.fill(0);
+    std::vector<std::array<size_t, 2>> path;
+    path.push_back({0, 0});
+    while (!path.empty())
+    {
+      auto node = path.back();
+      path.pop_back();
+
+      visited(node[0],node[1]) = 1;
+      int directions[4][2] = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
+      for (auto direction : directions)
+      {
+        size_t r_new = direction[0] + node[0];
+        size_t c_new = direction[1] + node[1];
+        if (!visited(r_new, c_new))
```

Comment:
I would say that the issue is coming from this line since it is the only line that is using `arma`.
You need to debug in this direction by printing the values and see when it might cause the issue.
I would not think that the issue is related to `azure` since all tests are failing

Reply:
When I change the board matrix to arma, the test fails at arma::Operator() (index out-of-bound), so I think the problem is when accessing the board indices.